tests/unit/test_signers.py

```diff
@@ -250,6 +250,63 @@ def test_presigned_url_throws_unsupported_signature_error(self):
             self.signer.generate_presigned_url({})
 
 
+class TestCloudfrontSigner(unittest.TestCase):
+    def setUp(self):
+        self.signer = CloudFrontSigner("MY_KEY_ID", lambda message: b'signed')
+
+    def test_b64encode(self):
+        self.assertEqual(b'dGVzdA__', self.signer._url_b64encode(b'test'))
+
+    def test_build_canned_policy(self):
+        policy = self.signer.build_policy('foo', datetime.datetime(2016, 1, 1))
+        expected = (
+            b'{"Statement":[{"Resource":"foo",'
+            b'"Condition":{"DateLessThan":{"AWS:EpochTime":1451606400}}}]}')
+        self.assertEqual(policy, expected)
+
+    def test_build_custom_policy(self):
+        policy = self.signer.build_policy(
+            'foo', datetime.datetime(2016, 1, 1),
+            date_greater_than=datetime.datetime(2015, 12, 1),
+            ip_address='12.34.56.78/9')
+        expected = (
+            b'{"Statement":['
+                b'{"Condition":{'
+                    b'"DateGreaterThan":{"AWS:EpochTime":1448928000},'
+                    b'"DateLessThan":{"AWS:EpochTime":1451606400},'
+                    b'"IpAddress":{"AWS:SourceIp":"12.34.56.78/9"}'
+                b'},'
+                b'"Resource":"foo"}'
+            b']}')
+        self.assertEqual(policy, expected)
+
+    def test_generate_presign_url_with_expire_time(self):
+        signed_url = self.signer.generate_presigned_url(
+            'http://test.com/foo.txt',
+            date_less_than=datetime.datetime(2016, 1, 1))
+        expected = (
+            'http://test.com/foo.txt?Expires=1451606400&Signature=c2lnbmVk'
+            '&Key-Pair-Id=MY_KEY_ID')
+        self.assertEqual(signed_url, expected)
+
+    def test_generate_presign_url_with_custom_policy(self):
+        policy = self.signer.build_policy(
+            'foo', datetime.datetime(2016, 1, 1),
+            date_greater_than=datetime.datetime(2015, 12, 1),
+            ip_address='12.34.56.78/9')
+        signed_url = self.signer.generate_presigned_url(
+            'http://test.com/index.html?foo=bar', policy=policy)
+        expected = (
```

Comment:
I would prefer we actually assert the various parts of the URL rather than an exact match, i.e parse the URL and assert the query string dict equals X, the hostname equals X, etc.  You'll get a much better diff on failure.